File: FibonacciFunction.py
import time
from functools import lru_cache
from typing import Optional

MAX_FIB_VALUE = 100000001  # Maximum value of n that can be calculated

# Memoization cache to store calculated Fibonacci values using LRU cache
@lru_cache(maxsize=None)
def fib_doubling(n: int) -> Optional[int]:
    # fib_doubling calculates the nth Fibonacci number using the doubling method
    if n < 2:
        return n  # Base cases: F(0) = 0, F(1) = 1
    elif n > MAX_FIB_VALUE:
        # Error if the value is too large to calculate within reasonable time
        raise ValueError(f"n is too large for this implementation. Please use a value less than or equal to {MAX_FIB_VALUE}.")
    # Calculate the Fibonacci value using an iterative helper function
    result = fib_doubling_helper_iterative(n)
    return result

# fib_doubling_helper_iterative is an iterative function that uses the doubling method to compute Fibonacci numbers
def fib_doubling_helper_iterative(n: int) -> int:
    # Initialize base Fibonacci values F(0) = 0 and F(1) = 1
    a, b = 0, 1

    # Determine the number of bits needed to represent n
    bit_length = n.bit_length()  # This helps in iterating efficiently over the binary representation of n

    # Iterate over each bit from the most significant to the least significant
    for i in reversed(range(bit_length)):
        # Use the doubling formulas to calculate F(2k) and F(2k + 1)
        # F(2k) = F(k) * [2 * F(k+1) - F(k)]
        c = a * ((b << 1) - a)  # Calculate F(2k) using F(k) and F(k+1)
        # F(2k + 1) = F(k)^2 + F(k+1)^2
        d = a * a + b * b  # Calculate F(2k + 1) using F(k) and F(k+1)

        # Update a and b based on the current bit of n
        if (n >> i) & 1 == 0:
            a, b = c, d  # If the bit is 0, set F(2k) to a and F(2k+1) to b
        else:
            a, b = d, c + d  # If the bit is 1, set F(2k+1) to a and F(2k+2) to b

    return a  # Return the nth Fibonacci number

# benchmark_fib benchmarks the Fibonacci calculations for a list of values
def benchmark_fib(n_values: list[int], repetitions: int):
    # Note: Clearing the cache ensures consistent results but may affect performance for repeated calculations.
    # Consider if this step is necessary based on your use case.

    for n in n_values:
        total_exec_time = 0
        min_exec_time = float('inf')  # Initialize min execution time to a very high value
        max_exec_time = float('-inf')  # Initialize max execution time to a very low value
        # Repeat the calculation for more accurate benchmarking
        for _ in range(repetitions):
            start = time.perf_counter()  # Start timing the calculation
            try:
                fib_doubling(n)  # Calculate the nth Fibonacci number
            except ValueError as err:
                # Handle the error if n is too large
                print(f"fib_doubling({n}): {err}")
                continue
            exec_time = (time.perf_counter() - start) * 1e9  # Convert seconds to nanoseconds
            total_exec_time += exec_time
            min_exec_time = min(min_exec_time, exec_time)  # Update min execution time if current is lower
            max_exec_time = max(max_exec_time, exec_time)  # Update max execution time if current is higher
        # Calculate the average execution time
        avg_exec_time = total_exec_time / repetitions
        # Print detailed benchmarking results including min, max, and average execution times
        print(f"fib_doubling({n}) averaged over {repetitions} runs: {avg_exec_time:.0f} nanoseconds (min: {min_exec_time:.0f} ns, max: {max_exec_time:.0f} ns)")

# main function to execute the benchmarking
def main():
    n_values = [1000000, 10000000, 100000000]  # List of values to benchmark
    repetitions = 3  # Number of repetitions for better accuracy
    benchmark_fib(n_values, repetitions)  # Run the benchmark

if __name__ == "__main__":
    main()

File: FibonacciFunction.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::Instant;
use std::cmp;
use num_bigint::BigInt;
use num_traits::{One, Zero};

const MAX_FIB_VALUE: usize = 100_000_001; // Maximum value of n that can be calculated
lazy_static::lazy_static! {
    static ref MEMO: Arc<Mutex<HashMap<usize, BigInt>>> = Arc::new(Mutex::new(HashMap::new()));
}

// fib_doubling calculates the nth Fibonacci number using the doubling method
fn fib_doubling(n: usize) -> Result<BigInt, String> {
    if n < 2 {
        return Ok(BigInt::from(n));
    } else if n > MAX_FIB_VALUE {
        return Err("n is too large for this implementation".to_string());
    }
    let result = fib_doubling_helper_iterative(n);
    Ok(result)
}

// fib_doubling_helper_iterative is an iterative function that uses the doubling method to compute Fibonacci numbers
fn fib_doubling_helper_iterative(n: usize) -> BigInt {
    let memo = MEMO.lock().unwrap();
    if let Some(val) = memo.get(&n) {
        return val.clone();
    }
    drop(memo);

    let mut a = BigInt::zero();
    let mut b = BigInt::one();

    let bit_length = 64 - n.leading_zeros() as usize;

    for i in (0..bit_length).rev() {
        let mut c = &b * 2u32 - &a;
        c = &a * &c;
        let d = &a * &a + &b * &b;

        if (n >> i) & 1 == 0 {
            a = c.clone();
            b = d.clone();
        } else {
            a = d.clone();
            b = c + d;
        }
    }

    let result = a.clone();
    let mut memo = MEMO.lock().unwrap();
    memo.insert(n, result.clone());
    result
}

// print_error prints an error message in a consistent format
fn print_error(n: usize, err: &str) {
    println!("fib_doubling({}): {}", n, err);
}

// benchmark_fib benchmarks the Fibonacci calculations for a list of values
fn benchmark_fib(n_values: &[usize], repetitions: usize) {
    let mut memo = MEMO.lock().unwrap();
    memo.clear();
    drop(memo);

    for &n in n_values {
        let mut total_exec_time = 0u128;
        for _ in 0..repetitions {
            let start = Instant::now();
            match fib_doubling(n) {
                Ok(_) => {
                    let elapsed = start.elapsed().as_nanos();
                    total_exec_time += elapsed;
                }
                Err(err) => {
                    print_error(n, &err);
                    continue;
                }
            }
        }
        let exec_time = total_exec_time / cmp::max(1, repetitions as u128);
        println!("fib_doubling({}) averaged over {} runs: {} nanoseconds", n, repetitions, exec_time);
    }
}

fn main() {
    let n_values = vec![1_000_, 10_000, 10_000_000];
    let repetitions = 3;
    benchmark_fib(&n_values, repetitions);
}
